kernel/src/transaction.rs

```diff
@@ -1,120 +1,249 @@
+            Err(crate::error::Error::ObjectStore(object_store::Error::AlreadyExists {
+                ..
+            })) => Ok(CommitResult::Conflict(self, commit_version)),
+            Err(e) => Err(e),
+        }
     }
 
-    /// Scan the Delta Log to obtain the latest transaction for all applications
-    pub fn application_transactions(&self, engine: &dyn Engine) -> DeltaResult<TransactionMap> {
-        self.scan_application_transactions(engine, None)
+    /// Add commit info to the transaction. This is commit-wide metadata that is written as the
+    /// first action in the commit. Note it is required in order to commit. If the engine does not
+    /// require any commit info, pass an empty `EngineData`.
+    pub fn commit_info(&mut self, commit_info: Box<dyn EngineData>, schema: Schema) {
+        self.commit_info = Some(EngineCommitInfo {
+            data: commit_info.into(),
+            schema: schema.into(),
+        });
     }
 }
 
-#[cfg(all(test, feature = "default-engine"))]
-mod tests {
-    use std::path::PathBuf;
+/// Result after committing a transaction. If 'committed', the version is the new version written
+/// to the log. If 'conflict', the transaction is returned so the caller can resolve the conflict
+/// (along with the version which conflicted).
+pub enum CommitResult {
+    /// The transaction was successfully committed at the version.
+    Committed(crate::Version),
+    /// The transaction conflicted with an existing version (at the version given).
+    Conflict(Transaction, crate::Version),
+}
+
+// given the engine's commit info (data and schema as [EngineCommitInfo]) we want to create both:
+// (1) the commitInfo action to commit (and append more actions to) and
+// (2) the schema for the actions to commit (this is determined by the engine's commit info schema)
+#[allow(clippy::type_complexity)]
+fn generate_commit_info(
+    engine: &'_ dyn Engine,
+    engine_commit_info: Option<EngineCommitInfo>,
+) -> DeltaResult<(
+    Box<dyn Iterator<Item = Box<dyn EngineData>> + Send + '_>,
+    SchemaRef,
+)> {
+    use crate::actions::{
+        ADD_NAME, COMMIT_INFO_NAME, METADATA_NAME, PROTOCOL_NAME, REMOVE_NAME, TRANSACTION_NAME,
+    };
+
+    // TODO: enforce single row commit info
+    // TODO: for now we always require commit info
+    let action_schema = Arc::new(engine_commit_info.as_ref().map_or(
+        get_log_schema().clone(),
+        |commit_info| {
+            let mut action_fields = get_log_schema().fields().collect::<Vec<_>>();
+            let commit_info_field = action_fields
+                .pop()
+                .expect("last field is commit_info in action schema");
+            let DataType::Struct(commit_info_schema) = commit_info_field.data_type() else {
+                unreachable!("commit_info_field is a struct");
+            };
+            let mut commit_info_fields = commit_info_schema.fields().collect::<Vec<_>>();
```

Comment:
Your call, this might be nicer to avoid materializing  and mutating vectors a bunch:
```rust
            let commit_info_fields = commit_info_schema
                .fields()
                .chain(commit_info.schema.fields())
                .cloned();
            let commit_info_schema = StructType::new(commit_info_fields.collect());
            let action_fields = action_fields.into_iter().cloned().chain(std::iter::once(
                crate::schema::StructField::new(COMMIT_INFO_NAME, commit_info_schema, true),
            ));
            StructType::new(action_fields.collect())
```

Reply:
nice!